packages/core-data/src/batch/create-batch.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * External dependencies
+ */
+import { isFunction, zip } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import defaultProcessor from './default-processor';
+
+/**
+ * Creates a batch, which can be used to combine multiple API requests into one
+ * API request using the WordPress batch processing API (/v1/batch).
+ *
+ * ```
+ * const batch = createBatch();
+ * const dunePromise = batch.add( {
+ *   path: '/v1/books',
+ *   method: 'POST',
+ *   data: { title: 'Dune' }
+ * } );
+ * const lotrPromise = batch.add( {
+ *   path: '/v1/books',
+ *   method: 'POST',
+ *   data: { title: 'Lord of the Rings' }
+ * } );
+ * const isSuccess = await batch.run(); // Sends one POST to /v1/batch.
+ * if ( isSuccess ) {
+ *   console.log(
+ *     'Saved two books:',
+ *     await dunePromise,
+ *     await lotrPromise
+ *   );
+ * }
+ * ```
+ *
+ * @param {Function} [processor] Processor function. Can be used to replace the
+ *                               default functionality which is to send an API
+ *                               request to /v1/batch. Is given an array of
+ *                               inputs and must return a promise that
+ *                               resolves to an array of objects containing
+ *                               either `output` or `error`.
+ */
+export default function createBatch( processor = defaultProcessor ) {
+	let lastId = 0;
+	let queue = [];
+	const pending = new ObservableSet();
+
+	return {
+		/**
+		 * Adds an input to the batch and returns a promise that is resolved or
+		 * rejected when the input is processed by `batch.run()`.
+		 *
+		 * You may also pass a thunk which allows inputs to be added
+		 * asychronously.
+		 *
+		 * ```
+		 * // Both are allowed:
+		 * batch.add( { path: '/v1/books', ... } );
+		 * batch.add( ( add ) => add( { path: '/v1/books', ... } ) );
+		 * ```
+		 *
+		 * If a thunk is passed, `batch.run()` will pause until either:
+		 *
+		 * - The thunk calls its `add` argument, or;
+		 * - The thunk returns a promise and that promise resolves, or;
+		 * - The thunk returns a non-promise.
```

Comment:
@noisysocks can you clarify the intention of the second and third options here? it seems to suggest that we can do this…

```ts
batch.add( () => Promise.resolve({ path: '/v1/books', ... }) );
```

or this…

```ts
batch.add( () => ({ path: '/v1/books', ... }) );
```

…however, if we do that, then `Promise.resolve(fetchOptionsProducer( queueForResolution))` will return the API request description and not the data requested in the batch operation.

- Am I misunderstanding the comment?
- Did we intend on eventually providing this functionality?
- Is this an un-noticed bug in the code?

I'm working on the types for this module and got stuck trying to understand the purpose here. I'm inclined to remove the second and third options since the only examples and existing core code only demonstrate the first.

Reply:
> …however, if we do that, then `Promise.resolve(fetchOptionsProducer( queueForResolution))` will return the API request description and not the data requested in the batch operation.

I'm not sure what you mean here. What's `fetchOptionsProducer`?

From memory this was all about preventing `run()` from pausing infinitely if `add()` is called with a thunk and that think never calls the function passed to it.

The comment is describing the effect of this code below:

```js
return Promise.resolve( inputOrThunk( add ) ).finally( () => {
	pending.delete( id );
} );
```

> - The thunk calls its `add` argument, or;

```js
const batch = createBatch();
batch.add( ( add ) => async {
	await sleep( 100 );
	add( {} ); // run() will resume here
	await sleep( 100 );
} );
batch.run(); // will pause for 100 ms
```

> - The thunk returns a promise and that promise resolves, or;

```js
const batch = createBatch();
batch.add( ( add ) => async {
	await sleep( 100 );
	await sleep( 100 );
	// run() will resume here when promise resolves
} );
batch.run(); // will pause for 200 ms
```

> - The thunk returns a non-promise.

```js
const batch = createBatch();
batch.add( ( add ) => {
	// run() will resume here as a non-promise is returned
} );
batch.run(); // will pause for 0 ms
```

> I'm working on the types for this module

I think it'd be something like this.

```typescript
interface Batch<TInput, TOutput> {
    add<TThunkResult>( inputOrThunk: TInput|( ( add: ( input: TInput ) => Promise<TOutput> ) => TThunkResult|Promise<TThunkResult> ) ): Promise<TOutput>|Promise<TThunkResult>;
}

function createBatch<TInput, TOutput>( processor: ( inputs: TInput[] ) => Promise<{output?: TOutput, error?: any}[]> ): Batch<TInput, TOutput> {
}
```

> I'm inclined to remove the second and third options since the only examples and existing core code only demonstrate the first.

Feel free to simplify! It's `__experimental` so should be okay to change. I believe the unit tests and usage in `@wordpress/core-data` cover everything we need `createBatch` to do.